### PR TITLE
refactor template cache

### PR DESCRIPTION
**Pull request by carlory**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

It solves the problem of this parallel test. 

 #78

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kwok/controllers/template_cache.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+	funcMap    template.FuncMap
+}
+
+func newTemplateCache(funcMap template.FuncMap) *templateCache {
+	return &templateCache{
+		funcMap: funcMap,
+		cache:   sync.Map{},
+		bufferPool: sync.Pool{
+			New: func() interface{} {
+				return new(bytes.Buffer)
```
Comment (wzshiming) [inline pkg/kwok/controllers/template_cache.go:42]:
```suggestion
				return bytes.NewBuffer(make([]byte, 4*1024))
```

Maybe it's better to allocate a little memory in advance

File: pkg/kwok/controllers/template_cache.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+			},
+		},
+	}
+}
+
+// renderToJson renders the template with the given text and original object.
+func (tc *templateCache) renderToJson(text string, original interface{}) ([]byte, error) {
```
Comment (wzshiming) [inline pkg/kwok/controllers/template_cache.go:49]: Can we add a unit test for this method

File: pkg/kwok/controllers/template_cache.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+	"text/template"
+
+	"sigs.k8s.io/yaml"
+)
+
+type templateCache struct {
```
Comment (wzshiming) [inline pkg/kwok/controllers/template_cache.go:30]: I think it would be better to call it Renderer

**Comment by carlory** [comment]
cc @wzshiming

File: pkg/kwok/controllers/renderer.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+		bufferPool: sync.Pool{
+			New: func() interface{} {
+				return bytes.NewBuffer(make([]byte, 4*1024))
+			},
+		},
+	}
+}
+
+// renderToJson renders the template with the given text and original object.
+func (tc *renderer) renderToJson(text string, original interface{}) ([]byte, error) {
```
Comment (wzshiming) [inline pkg/kwok/controllers/renderer.go:49]:
```suggestion
func (r *renderer) renderToJson(text string, original interface{}) ([]byte, error) {
```

**Pull request merged**